### PR TITLE
Create_dataset HTTP API

### PR DESCRIPTION
**Pull request by cecilia-uu**
### What problem does this PR solve?

Add create_dataset HTTP API.

### Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)

### CODE REVIEW

File: api/apps/__init__.py [hunk 1]
```diff
@@ -70,6 +70,7 @@ def register_page(page_path):
     page_name = page_path.stem.rstrip('_app')
     module_name = '.'.join(page_path.parts[page_path.parts.index('api'):-1] + (page_name, ))
 
+    print(page_name, module_name)
```
Comment (JinHai-CN) [inline api/apps/__init__.py]: Remove log?

File: api/apps/__init__.py [hunk 2]
```diff
@@ -79,6 +80,8 @@ def register_page(page_path):
 
     page_name = getattr(page, 'page_name', page_name)
     url_prefix = f'/{API_VERSION}/{page_name}'
+    if page_name == 'sdk':
```
Comment (JinHai-CN) [inline api/apps/__init__.py:85]: not only for sdk, also valid for other page_name.

File: api/apps/sdk_app.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    objs = APIToken.query(token=token)
+
+    # Authorization error
+    if not objs:
+        return get_json_result(
+            data=False, retmsg='Token is not valid!"', retcode=RetCode.AUTHENTICATION_ERROR)
+
+    tenant_id = objs[0].tenant_id
+    req = request.json
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: use full name: request_body as the var name.

File: api/apps/sdk_app.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        token = request.headers.get('Authorization').split()[1]
+    except IndexError:
+        return get_json_result(data=False, retmsg="Token is missing!", retcode=RetCode.AUTHENTICATION_ERROR)
+
+    objs = APIToken.query(token=token)
+
+    # Authorization error
+    if not objs:
+        return get_json_result(
+            data=False, retmsg='Token is not valid!"', retcode=RetCode.AUTHENTICATION_ERROR)
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: retmsg => message
retcode => code

File: api/apps/sdk_app.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        req["id"] = get_uuid()
+        req["tenant_id"] = tenant_id
+        req["created_by"] = tenant_id
+        e, t = TenantService.get_by_id(tenant_id)
+        if not e:
+            return get_data_error_result(retmsg="Tenant not found.")
+        req["embd_id"] = t.embd_id
+        if not KnowledgebaseService.save(**req):
+            return get_data_error_result()
+        return get_json_result(data={"kb_id": req["id"]})
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: kb_id => dataset_id

File: api/apps/sdk_app.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        tenant_id=tenant_id,
+        status=StatusEnum.VALID.value)
+    try:
+        req["id"] = get_uuid()
+        req["tenant_id"] = tenant_id
+        req["created_by"] = tenant_id
+        e, t = TenantService.get_by_id(tenant_id)
+        if not e:
+            return get_data_error_result(retmsg="Tenant not found.")
+        req["embd_id"] = t.embd_id
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: What is embd_id ? use full name.

File: api/apps/sdk_app.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    if not req["name"]:
+        return get_data_error_result(retmsg="Name cannot be empty!")
+
+    # In case that the length of the name exceeds the limit
+    if len(req["name"]) > LIMIT_LENGTH_OF_NAME:
+        return get_data_error_result(retmsg=f"The name of the dataset exceeds {LIMIT_LENGTH_OF_NAME}!")
+
+    # In case that there is space in the name
+    if contains_space(req["name"]):
+        return get_data_error_result(retmsg="There is space in this dataset's name. Please remove it.")
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: Aren't we allow the space in name?

File: api/apps/sdk_app.py
Comment (JinHai-CN) [inline api/apps/sdk_app.py:1]: File name: dataset.py

File: api/apps/sdk_app.py (hunk 4 above)
Comment (JinHai-CN) [reply api/apps/sdk_app.py]: Message: Token is invalid.

File: api/apps/sdk_app.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+    tenant_id = objs[0].tenant_id
+    req = request.json
+
+    # In case that there's space in the end or the start
+    req["name"] = req["name"].strip()
+
+    # In case that there's no name
+    if not req["name"]:
+        return get_data_error_result(retmsg="Name cannot be empty!")
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: 'name' field is required.

File: api/apps/sdk_app.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            data=False, retmsg='Token is not valid!"', retcode=RetCode.AUTHENTICATION_ERROR)
+
+    tenant_id = objs[0].tenant_id
+    req = request.json
+
+    # In case that there's space in the end or the start
+    req["name"] = req["name"].strip()
+
+    # In case that there's no name
+    if not req["name"]:
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: Tell the 'name' field existence before strip the 'name' field value.

File: api/apps/sdk_app.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    # Authorization error
+    if not objs:
+        return get_json_result(
+            data=False, retmsg='Token is not valid!"', retcode=RetCode.AUTHENTICATION_ERROR)
+
+    tenant_id = objs[0].tenant_id
+    req = request.json
+
+    # In case that there's space in the end or the start
+    req["name"] = req["name"].strip()
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: Use another var to store the value, to avoid performance downgrade.

File: api/apps/sdk_app.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    # In case that the length of the name exceeds the limit
+    if len(req["name"]) > LIMIT_LENGTH_OF_NAME:
+        return get_data_error_result(retmsg=f"The name of the dataset exceeds {LIMIT_LENGTH_OF_NAME}!")
+
+    # In case that there is space in the name
+    if contains_space(req["name"]):
+        return get_data_error_result(retmsg="There is space in this dataset's name. Please remove it.")
+
+    # In case that there are other fields in the data-binary
+    if len(req.keys()) > 1:
```
Comment (JinHai-CN) [inline api/apps/sdk_app.py]: Please also mention which fields are not allowed.

**Comment by JinHai-CN** [comment]
What is the test code?

File: api/apps/__init__.py (hunk 2 above)
Comment (JinHai-CN) [reply api/apps/__init__.py:85]: How to implement other API in other name file?

File: api/apps/__init__.py (hunk 1 above)
Comment (cecilia-uu) [reply api/apps/__init__.py]: Done

File: api/apps/sdk_app.py (hunk 3 above)
Comment (cecilia-uu) [reply api/apps/sdk_app.py]: Done

File: api/apps/sdk_app.py (hunk 4 above)
Comment (cecilia-uu) [reply api/apps/sdk_app.py]: Done

File: api/apps/sdk_app.py (hunk 5 above)
Comment (cecilia-uu) [reply api/apps/sdk_app.py]: Done

File: api/apps/sdk_app.py (hunk 6 above)
Comment (cecilia-uu) [reply api/apps/sdk_app.py]: It means embedding ID, which is stored in the database.

File: api/apps/sdk_app.py
Comment (cecilia-uu) [reply api/apps/sdk_app.py:1]: Done.

**Pull request closed without merging**